src/botPage/view/blockly/customBlockly.js

```diff
@@ -376,3 +377,263 @@ Blockly.WorkspaceAudio.prototype.preload = function() {
         }
     }
 };
+
+// https://groups.google.com/forum/#!msg/blockly/eS1V49pI9c8/VEh5UuUcBAAJ
+// Custom Variable Block with Download Context Menu
+Blockly.defineBlocksWithJsonArray([
```

Comment:
Re-defining the blocks is a more correct way, however, the definitions should be within their own files and use the same code style that we use everywhere (i.e. we never use `defineBlocksWithJsonArray` and use ES6 where possible).

Also, we don't necessarily need to redefine the blocks, instead, we can override both mixins `customContextMenu` functions in `customBlockly` to achieve the same behaviour. See example below.


```JavaScript
import { save } from './utils';

const addDownloadOption = (callback, options) => {
    options.push({
        text    : translate('Download'),
        enabled : true,
        callback: () => {
            const xml = Blockly.Xml.textToDom(
                '<xml xmlns="http://www.w3.org/1999/xhtml" collection="false"></xml>'
            );
            xml.appendChild(Blockly.Xml.blockToDom(this));
            save('binary-bot-block', true, xml);
        },
    });
    callback(options);
};

const originalCustomContextVarFn = Blockly.Constants.Variables.CUSTOM_CONTEXT_MENU_VARIABLE_GETTER_SETTER_MIXIN.customContextMenu;
Blockly.Constants.Variables.CUSTOM_CONTEXT_MENU_VARIABLE_GETTER_SETTER_MIXIN.customContextMenu = function(options) {
    addDownloadOption(originalCustomContextVarFn.bind(this), options);
}

const originalCustomContextLoopFn = Blockly.Constants.Loops.CUSTOM_CONTEXT_MENU_CREATE_VARIABLES_GET_MIXIN.customContextMenu;
Blockly.Constants.Loops.CUSTOM_CONTEXT_MENU_CREATE_VARIABLES_GET_MIXIN.customContextMenu = function(options) {
    addDownloadOption(originalCustomContextLoopFn.bind(this), options);
}
```

Reply:
genius!